### ISSUE TITLE
DomainIntegrator on a region of the domain

### ISSUE THREAD

**Issue by tomathew** [mfem/mfem#877]

**Comment by tomathew** [comment]
Sorry ! 
created by accident.
Regards,
Tom

**Comment by tomathew** [comment]
The original query (that I closed after finding aobut VectorRestrictedCoefficient) was about defining VectorFunctionCoefficients on subsets of domain marked by domain attributes.
I found just after posting the subject, that VectorRestrictedCoefficient can be used to define VectorFunctionCoefficients on subsets of the domain or biundaries. Apologizing the inconvenience.
I have some more queries related to this:
(1). I am working on very complex micro-scale geometries (mostly obtained from X-tomography stl files), so is this VectorRestrictedCoefficient robust enough to work all the time. As the coefficient is evaluated outside in a function, where the mesh is not available, I couldn't make coefficients zero where it shouldn't be, explicitly. Is there a work around ?

(2). I am little confused between mesh.attributes.Max or mesh.attributes.Size to be used. Sometimes the triangulation I have to work on come with domain attributes starting at 110 to 150, say if there are 5 regions it is {110,120,130,140,150}. Do I have to change it to {1,2,3,4,5} ?

Best regards,
Tom

**Comment by tomathew** [comment]
Hello again,
To add to the queries above,
like AddBoundaryIntegrator() overloaded with an argument for bdr_attributes, does a similar kind of overload exist for DomainIntegrator() in any of the branches of MFEM ? 
Best regards,
Tom
P.S: can't help but say that I am enjoying the MFEM platform, especially the way the elemental mathematical objects of the finite element framework is perceived and implemented. It is giving me a different perspective than the platform I was using previously, though it is little difficult for me at this initial stage to understand this wonderful platform. Kudos !

**Comment by mlstowell** [comment]
Hi, @tomathew ,
Unfortunately I don't fully understand your first query.  `VectorRestrictedCoefficient` should always work as it was designed to do but that design may not be adequate for your requirements.  In particular I doubt if many of our `Coefficient` classes would return meaningful data without a mesh.  They are evaluated using an `ElementTransformation` which provides, among other things, the attribute number.  You might be able to rig up a dummy `ElementTransformation` which contains an attribute number that you assign to return zero and hence avoid any calls to the mesh.  However, coefficients are usually called from within loops over mesh elements so it's not clear to me how you would call this where the mesh is not available.

Your second question is more straightforward.  We want to use the attribute numbers as indices in arrays because array look-ups are faster than working with std::maps or other types of hash tables.  For this reason we often use arrays that are indexed from `0` to `mesh.attributes.Max-1`.  There is no requirement that all of these numbers be used or that your attribute numbers be contiguous.  Our only requirement is that all attribute numbers be positive integers.  This can potentially be wasteful if your maximum attribute number is very large.  How much memory you're willing to use for this purpose is up to you.  A few hundred extra integers is probably not going to be noticeable but if you're concerned you can define your own mapping and renumber them (although MFEM does not provide any quick means to accomplish this that I'm aware of.)

I don't recall seeing any branches that use attribute marker arrays with `AddDomainIntegrator` methods.  This would, however, be a useful addition.  We should consider adding it.

Best wishes,
Mark

**Comment by tomathew** [comment]
Hello Mark, 
Thanks for the reply. Sorry for complicating things. The weak formulation I am trying to implement looks like this below:
![varf](https://user-images.githubusercontent.com/22793567/56473377-e4d53600-646a-11e9-8e0b-b78075716b09.png)
As you can see the linear form of the linear system is defined on a subset of the domain. The original query (first question) was when I didn't know about VectorRestrictedCoefficient(). But later I found out that option and it seems to be working fine. The point about mesh info being not available for (mesh.attribute()) was in the function for the incident field Ei, which is for ex: 
void Einc(const Vector &x, Vector &Ei);  in this one. Anyway that seems to be not a problem for me now, as I use VectorRestrictedCoefficient().

After changing the mesh attributes in my input triangulation to {1,2,3,4,5}, I could roughly setup my basic solver. But I am afraid of new concerns or the way I implemented it. It is compiling and running. As you can see in the figure the linear form involves a product of PWConstantCoefficient(epsilon) and the incident field. The way I implemented this is first by projecting all these to a GridFunction, multiply these coefficients and later setting it to a VectorGridFunction() by the SetGridFunction(). I am not sure if this is the optimal way. Is there a better way ? (I'll take a look at the hertz_solver.cpp, you suggested, tomorrow. First I was trying to setup a basic running version of my solver)

With that being said, after facing some memory issues, GridFunction for the PWConstCoefficients() are defined on a fespace of H1_FECollection(pOrder+1), where pOrder is the Nedelec basis order where Ei lives. On using pOrder=Nedelec order, the linear form was evaluated to be zero. This framework works fine in serial.

In parallel, for a start I am trying to solver the blockOperator with BICGSTAB (implemented as per the parallel Darcy problem), but these are some issues, which are listed below:
mpirun -np 1 ./exe , work fine without any memory or pointer error, but on multiple processors I am having some issues. When using pOrder+1 for the PWConstCoefficient(), the residual starts at -NAN (I have no idea now !), which is fixed by changing to pOrder+2.
On multiple processors, the linear system finishes the solving, but throws some memory pointer exceptions at the end, which I couldn't look in detail yet (I'll look more closely and get back). May be I have to see the examples in more detail. This is just letting you know the present situation. 
It's going well, I am very much looking forward to work on this platform.

Lastly, after seeing the weak form (in rough), if you have some directions, suggestions for solver tuning or preconditioning, please let me know. I am a PhD student. All suggestions are welcomed and will be appreciated. Now my plan is to give AMS a try.
I hope everything is clear, else let me know. Apologizing this lengthy thread.
Thankfully,
Tom

**Comment by tomathew** [comment]
Hello again.
Now it's clear the first part your reply, I will see that soon. I was yesterday before re-opening the issure going through the other related threads, and ended up somewhere similar issue was raised. With that in my mind, I posted the above query. Of course, I will surely give this a try. Also, VectorRestricedCoefficient() seems to be working fine for me now. So this is good !
Thanks again,
Tom

**Comment by mlstowell** [comment]
Hi, @tomathew ,
I happened to notice a small typo in your equations. In the equation for L^i I believe you should have a positive sign before the term involving U^r_{inc}.

The best scheme for handling coefficients depends on how the coefficients are computed to begin with.  I would not project epsilon and U_inc onto `ParGridFunction` objects unless there is a compelling reason.  If have analytic, or piece-wise analytic, functions for these quantities then it would be best to combine them either as functions or as `Coefficients`.  For example, assume you have functions for `epsilon_sc`, `epsilon_h`, and `U_inc` (I'll ignore real and imaginary parts).  The most efficient code would come from writing an other function which calls these three and computes `(epsilon_sc-epsilon_h)*U_inc`.  You then create a `VectorFunctionCoefficient` which calls this function and pass it to the linear form.

An alternative is to combine the functions as `Coefficients`.  This will be less efficient but may be more convenient.  In this situation you create `Coefficients` to represent `epsilon_sc` and `epsilon_h` and pass these to a `SumCoefficient` object to compute their difference.  Then create a `VectorCoefficient` to represent `U_inc` and use the `ScalarVectorProductCoefficient` to multiply this by the difference of `epsilon` coefficients.

The thing you never want to do is multiply `GridFunction` objects together entry by entry.  There are special situations where this works as you might expect but in the vast majority of cases such a product is meaningless.  I apologize if this seems obvious but it sounds like you may have tried this.

Regarding solvers you may be interested in a paper which discusses using [AMS on a similar equation](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=2ahUKEwi-5p7h7uHhAhUNbq0KHUgWD5UQFjAAegQIBhAB&url=https%3A%2F%2Flibrary.seg.org%2Fdoi%2F10.1190%2Fgeo2015-0013.1&usg=AOvVaw3V0qYXNzkUMBZ90Ms8ZedN).

Best wishes,
Mark

P.S. I'll be glad to offer more feedback as you work on this.

**Comment by mlstowell** [comment]
Oh, one note regarding the iterative solver.  If you multiply the second row of your block system by -1 the matrix will become symmetric.  This will allow you to use MINRES which may be more efficient than BiCGSTAB.  In either case the crucial part will be the preconditioner (see the paper mentioned in the last post).

**Comment by tomathew** [comment]
Dear Mark,
This was an excerpt from my first year report. I am in my second year. And, my existing solver is in a platform which support complex algebra. So, I have never depended on this varf before. I'll definitely take a look soon and get back to you. Please note that by the sign convention I follow, complex electrical permittivity \epsilon = \epsilon^\prime + 1i * \epsilon^{\prime\prime}, which means I have + ve sign. I am from a Mechanical Engineering background, so not an expert on Electromagnetic theory. But, I think the electrical community follow a different sign convention (- ve ), which can be your case, I am not sure. 
I'll definitely take a look closer.

**Comment by tomathew** [comment]
I'll definitely see your comments regarding coefficients, and improve my solver. Thank you for the reference, the answers, etc. I will take some time to see if there is a mistake too. I recall validating this varf with my complex varf in my first year, so I don't think there is an issue. But still, I will do a recheck on this. Thank you.
Thankfully,
Tom

**Comment by mlstowell** [comment]
Hi, @tomathew 
The negative sign between the two main terms in the expression for `L^r` comes from the `i^2` that arises when you multiply the imaginary part of `epsilon` by the imaginary part of `U_inc`.  On the other hand, when computing `L^i` there are no `i^2` factors so both terms in `L^i` will have the same sign.

Best wishes,
Mark

**Comment by tomathew** [comment]
Hello Mark,
Okay, I'll check it.
Thanks,
Tom

**Comment by tomathew** [comment]
Dear Mark,
Thanks you very much, as I could implement the Coefficents in a more elegant way. In fact I was unaware of the functions SumCoefficients() and ScalarVectorProductCoefficient(). I followed the second methodology you explained in https://github.com/mfem/mfem/issues/877#issuecomment-485274608 . 
Are there any examples closer or similar to the first methodology you explained in https://github.com/mfem/mfem/issues/877#issuecomment-485274608 ? 
I am aware of example 17. My doubt regarding first approach is, epsilon is a PWConstCoefficient in space, f(x), so I have to use ScalarVectorProductCoefficient at some point. 
Lastly, are there are other resources than the code documentation, where I can find or be aware of the available functions in mfem (any cheatsheets or pdf or technical reports which is worth reading ) ?
Then, Who is responsible for closing issues ? Is it the author or you ? I am new to these "issue" !
Thankfully,
Tom
P.S: I think I passed the initial and difficult stage on a new platform, as I setup successfully my solver in serial and parallel. Now the numerical experiments can begin. See you in another query !

**Comment by mlstowell** [comment]
Hi, @tomathew ,
One example of the first approach is shown in [cold plasma `DielectricTensor` class](https://github.com/mfem/mfem/blob/09400a4edbd3c2d343ebaf666340a07188b4dfdf/miniapps/plasma/cold_plasma_dielectric_coefs.cpp#L235).  However, you make a good point about the dependence on the attribute number in your coefficients and the restrictions this creates.  It sounds like you've chosen the best solution.

Currently MFEM does not have comprehensive documentation.  Even the header files don't adequately document many of the classes and methods.  The only other place I can point to is our web site [MFEM.org](https://mfem.org).  The "Documentation" link at the top of this page leads to several documents that we hope to continuously expand as needed.

Best wishes,
Mark

**Comment by tzanio** [comment]
Regarding documentation -- we are always open to help from our users :)

Our main focus so far has been on example codes and miniapps, which we try to keep updated with the latest features.  If there is an important use case not illustrated there, please feel free to suggest a modification to one of the existing examples/miniapps, or propose a new one.

What we are missing are more detailed tutorials, something like https://mfem.org/serial-tutorial/ and https://xsdk-project.github.io/ATPESC2018HandsOnLessons/lessons/mfem_convergence/ but with more details. 

If anyone in @mfem/users is willing to contribute such tutorials, e.g. based on their "what would have been helpful to know when I started" experience, we will be really thankful for that! Such tutorials can be proposed as simple Markdown files in the https://github.com/mfem/web repository, see e.g. https://raw.githubusercontent.com/mfem/web/master/src/serial-tutorial.md.

**Comment by mlstowell** [comment]
Hi, @tomathew, 
I forgot to address the "issue closing" issue.  There are no specific rules that I'm aware of.  Typically we see the person who opened the issue will close it once they are satisfied with the response or have no further questions.  If we notice issues that sit idle for a time we will occasionally close them if we think a resolution has been reached.  People are always welcome to re-open an issue if more questions arise.

Best wishes,
Mark

**Comment by tomathew** [comment]
Dear @tzanio , @mlstowell ,
Thanks for the reply ! 
@tzanio Thanks for sharing the link for hands-on tutorials. I'll go through it, and take time to get familiar with the platform. I would be happy to share, if I could develop something good. I prefer to take time to understand the platform, and develop something good. We will see in time.
@mlstowell  Thanks ! I will go  through the examples, hertz.cpp and plasma one, and others, to understand this platform, and get back to this mailing list if I am stuck somewhere.

I think the .msh reader is not accepting meshes (.msh) from a re-meshing software https://www.mmgtools.org/ . I am getting this error: 

MFEM abort: Non-positive element attribute in Gmsh mesh!
 ... in function: void mfem::Mesh::ReadGmshMesh(std::istream&)
 ... in file: mesh/mesh_readers.cpp:1172

Aborted (core dumped)
____________________
I'll close this issue, and open another one, after confirming, may I can share sample mesh files too.
Thankfully,
Tom

**Comment by tomathew** [comment]
Hello again,
Do you prefer me creating a new issue, or link the aforementioned issue in https://github.com/mfem/mfem/issues/678#issue-381996213 ?
Regards
Tom

**Comment by tzanio** [comment]
Hi Tom,

Please create a new issue for this. Is the output of mmgtools in [Gmsh](http://gmsh.info/) format, or its own format?

Tzanio
